File: com-mini-core/src/main/java/com/mini/core/security/crypto/BaseMac.java
package com.mini.core.security.crypto;

import com.mini.core.security.SecurityBase;

import javax.crypto.Mac;
import javax.crypto.spec.SecretKeySpec;
import java.nio.ByteBuffer;
import java.nio.charset.Charset;
import java.security.spec.AlgorithmParameterSpec;

import static javax.crypto.Mac.getInstance;

public abstract class BaseMac extends SecurityBase<BaseMac> {
    private final String algorithm;
    private final Mac mac;

    public BaseMac(String algorithm) throws Exception {
        this.mac       = getInstance(algorithm);
        this.algorithm = algorithm;
    }

    public final BaseMac init(byte[] input, int offset, int len, AlgorithmParameterSpec params) throws Exception {
        mac.init(new SecretKeySpec(input, offset, len, algorithm), params);
        return this;
    }

    public final BaseMac init(byte[] input, int offset, int len) throws Exception {
        mac.init(new SecretKeySpec(input, offset, len, algorithm));
        return this;
    }

    public final BaseMac init(byte[] input, AlgorithmParameterSpec params) throws Exception {
        mac.init(new SecretKeySpec(input, algorithm), params);
        return this;
    }

    public final BaseMac init(byte[] input) throws Exception {
        mac.init(new SecretKeySpec(input, algorithm));
        return this;
    }

    public final BaseMac init(String input, Charset charset, AlgorithmParameterSpec params) throws Exception {
        return init(input.getBytes(charset), params);
    }

    public final BaseMac init(String input, Charset charset) throws Exception {
        return init(input.getBytes(charset));
    }

    public final BaseMac init(String input, AlgorithmParameterSpec params) throws Exception {
        return init(input.getBytes(), params);
    }

    public final BaseMac init(String input) throws Exception {
        return init(input.getBytes());
    }

    public final BaseMac update(byte[] input, int offset, int len) {
        mac.update(input, offset, len);
        return this;
    }

    public final BaseMac update(byte[] input) {
        mac.update(input);
        return this;
    }

    public final BaseMac update(byte input) {
        mac.update(input);
        return this;
    }

    public final BaseMac update(ByteBuffer input) {
        mac.update(input);
        return this;
    }

    public final BaseMac doFinal(byte[] output, int outOffset) throws Exception {
        mac.doFinal(output, outOffset);
        return this;
    }

    public final byte[] doFinal(byte[] input) {
        return mac.doFinal(input);
    }

    public final byte[] doFinal() {
        return mac.doFinal();
    }

    @Override
    public byte[] digest(byte[] input) {
        return doFinal(input);
    }

    @Override
    public byte[] digest() {
        return doFinal();
    }

    public final String toString() {
        return mac.toString();
    }

    public final void reset() {
        mac.reset();
    }
}

File: mini-util/src/main/kotlin/com/mini/core/security/crypto/BaseMac.kt
@file:Suppress("unused", "MemberVisibilityCanBePrivate")

package com.mini.core.security.crypto

import com.mini.core.security.SecurityBase
import java.nio.ByteBuffer
import java.nio.charset.Charset
import java.security.spec.AlgorithmParameterSpec
import javax.crypto.Mac
import javax.crypto.spec.SecretKeySpec

abstract class BaseMac constructor(private val algorithm: String) : SecurityBase<BaseMac>() {
    private var mac: Mac = Mac.getInstance(algorithm)

    fun init(input: ByteArray?, offset: Int, len: Int, params: AlgorithmParameterSpec?): BaseMac {
        mac.init(SecretKeySpec(input, offset, len, algorithm), params)
        return this
    }

    fun init(input: ByteArray?, offset: Int, len: Int): BaseMac {
        mac.init(SecretKeySpec(input, offset, len, algorithm))
        return this
    }

    fun init(input: ByteArray?, params: AlgorithmParameterSpec?): BaseMac {
        mac.init(SecretKeySpec(input, algorithm), params)
        return this
    }

    fun init(input: ByteArray?): BaseMac {
        mac.init(SecretKeySpec(input, algorithm))
        return this
    }

    fun init(input: String, charset: Charset?, params: AlgorithmParameterSpec?): BaseMac {
        return init(input.toByteArray(charset!!), params)
    }

    fun init(input: String, charset: Charset?): BaseMac {
        return init(input.toByteArray(charset!!))
    }

    fun init(input: String, params: AlgorithmParameterSpec?): BaseMac {
        return init(input.toByteArray(), params)
    }

    fun init(input: String): BaseMac {
        return init(input.toByteArray())
    }

    override fun update(input: ByteArray, offset: Int, len: Int): BaseMac {
        mac.update(input, offset, len)
        return this
    }

    override fun update(input: ByteArray): BaseMac {
        mac.update(input)
        return this
    }

    override fun update(input: Byte): BaseMac {
        mac.update(input)
        return this
    }

    fun update(input: ByteBuffer?): BaseMac {
        mac.update(input)
        return this
    }

    fun doFinal(output: ByteArray?, outOffset: Int): BaseMac {
        mac.doFinal(output, outOffset)
        return this
    }

    fun doFinal(input: ByteArray?): ByteArray {
        return mac.doFinal(input)
    }

    fun doFinal(): ByteArray {
        return mac.doFinal()
    }

    override fun digest(input: ByteArray): ByteArray {
        return doFinal(input)
    }

    override fun digest(): ByteArray {
        return doFinal()
    }

    override fun toString(): String {
        return mac.toString()
    }

    fun reset() {
        mac.reset()
    }

}